rfc/src/rfcs/0002-function-stubbing.md

```diff
@@ -0,0 +1,416 @@
+
+1. a specified `replacement` stub does not exist;
+2. the user specifies conflicting stubs for the same harness (i.e., if the same `original` function is mapped to multiple `replacement` functions); or
+3. the signature of the `replacement` function is not compatible with the signature of the `original` function.
+
+### Pedagogy
+
+To teach this feature, we will update the documentation with a section on function and method stubbing, including simple examples showing how stubbing can help Kani handle code that currently cannot be verified.
+
+## Detailed Design
+
+We expect that this feature will require changes primarily to `kani-compiler`.
+Before doing code generation, `kani-compiler` already collects harnesses; we will extend this to also collect stub mapping information.
+Since stubs are specified on a per-harness basis, we need to generate multiple versions of code if all harnesses do not agree on their stub mappings; accordingly, we will update `kani-compiler` to generate multiple versions of code as appropriate. 
+To do the stubbing, we will plug in a new MIR-to-MIR transformation that replaces the bodies of specified functions with their replacements.
+This can be achieved via `rustc`'s query mechanism: if the user wants to replace `foo` with `bar`, then when the compiler requests the MIR for `foo`, we instead return the MIR for `bar`.
+`kani-compiler` will also be responsible for checking for the error conditions previously enumerated.
+
+We anticipate that this design will evolve and be iterated upon.
+
+## Rationale and alternatives: user experience
+
+Stubbing is a *de facto* necessity for verification tools, and the lack of stubbing has a negative impact on the usability of Kani.
+
+### Benefits
+
+- Because stubs are specified by annotating the harness, the user is able to specify stubs for functions they do not have source access to (like library functions).
+This contrasts with annotating the function to be replaced (such as with function contracts).
+- The current design provides the user with flexibility, as they can specify different sets of stubs to use for different harnesses.
+This is important if users are trying to perform compositional reasoning using stubbing, since in some harnesses a function/method should be verified, in in other harnesses its behavior should be assumed.
+- The stub selections are located adjacent to the harness, which makes it easy to understand which replacements are going to happen for each harness.
+
+### Risks
+
+- Users can always write stubs that do not correctly correspond to program behavior, and so a successful verification does not actually mean the program is bug-free.
+This is similar to other specification bugs. 
+
+### Comparison to function contracts
+
+- In many cases, stubs are more user-friendly than contracts. With contracts, it is necessary to explicitly provide information that is automatically captured in Rust (such as which memory is written).
```

Comment:
Are `assign` clauses really necessary? A good point is that contract predicates are a DSL on its own... Not just Rust code.